Automattic/wp-calypso
# Checkout: Save contact details when editing after validation

#### Changes proposed in this Pull Request

When you progress through checkout, one of the steps is the contact/billing information form. The contents of this form varies based on what products are in your cart, showing different fields for domain registrations, email products, and plans. Often, these fields will be pre-filled; the data to fill them comes from the `/me/domain-contact-information` endpoint and is updated by the `/me/transactions` endpoint when the checkout form has been submitted. However, if the transaction is not submitted or if it fails, any edited contact information in the form will not be saved to the server. This can be annoying if, for example, you need to reload checkout for some reason and are then forced to re-enter your contact details.

In this PR we modify the checkout contact information step to also send the edited contact details to the server via a new POST version of the `/me/domain-contact-information` endpoint added by D78812-code when validation of the details is complete. This way they will be preserved even if the page is reloaded or if the transaction fails. 

Fixes https://github.com/Automattic/wp-calypso/issues/43265

#### Testing instructions

- Apply D78812-code and sandbox the API.
- Add a product to your cart and visit checkout (we should test this twice, once with a plan and once with a domain product to get different forms).
- Click to edit your contact/billing details (if the step has been auto-completed).
- Change some of the contact details and click "Continue".
- Once the step is completed, reload the page and verify that the changes you made have persisted.
